Source/Core/Core/HW/WiimoteEmu/Encryption.cpp

```diff
@@ -6,281 +6,503 @@
+  ft[7] = sboxes[idx][rand[1] + offset] ^ sboxes[idx + 1][rand[8] + offset];
 
-  sb[0] = sboxes[idx][key[0]] ^ sboxes[(idx + 1) % 8][rand[1]];
-  sb[1] = sboxes[idx][key[5]] ^ sboxes[(idx + 1) % 8][rand[4]];
-  sb[2] = sboxes[idx][key[3]] ^ sboxes[(idx + 1) % 8][rand[0]];
-  sb[3] = sboxes[idx][key[2]] ^ sboxes[(idx + 1) % 8][rand[9]];
-  sb[4] = sboxes[idx][key[4]] ^ sboxes[(idx + 1) % 8][rand[7]];
-  sb[5] = sboxes[idx][key[1]] ^ sboxes[(idx + 1) % 8][rand[8]];
-  sb[6] = sboxes[idx][rand[3]] ^ sboxes[(idx + 1) % 8][rand[5]];
-  sb[7] = sboxes[idx][rand[2]] ^ sboxes[(idx + 1) % 8][rand[6]];
+  sb[0] = sboxes[idx][key[0] + offset] ^ sboxes[idx + 1][rand[1] + offset];
+  sb[1] = sboxes[idx][key[5] + offset] ^ sboxes[idx + 1][rand[4] + offset];
+  sb[2] = sboxes[idx][key[3] + offset] ^ sboxes[idx + 1][rand[0] + offset];
+  sb[3] = sboxes[idx][key[2] + offset] ^ sboxes[idx + 1][rand[9] + offset];
+  sb[4] = sboxes[idx][key[4] + offset] ^ sboxes[idx + 1][rand[7] + offset];
+  sb[5] = sboxes[idx][key[1] + offset] ^ sboxes[idx + 1][rand[8] + offset];
+  sb[6] = sboxes[idx][rand[3] + offset] ^ sboxes[idx + 1][rand[5] + offset];
+  sb[7] = sboxes[idx][rand[2] + offset] ^ sboxes[idx + 1][rand[6] + offset];
 }
 
-namespace WiimoteEmu
+// Generate key from the 0x40-0x4f data in the extension register.
+EncryptionKey KeyGen::GenerateFromExtensionKeyData(const std::array<u8, 0x10>& key_data) const
 {
-// Generate key from the 0x40-0x4c data in g_RegExt
-void EncryptionKey::Generate(const u8* const keydata)
-{
-  u8 rand[10];
-  u8 skey[6];
-  u8 testkey[6];
-  int idx;
+  std::array<u8, 10> rand;
+  for (int i = 0; i != 10; ++i)
+    rand[9 - i] = key_data[i];
 
-  for (int i = 0; i < 10; ++i)
-    rand[9 - i] = keydata[i];
-  for (int i = 0; i < 6; ++i)
-    skey[5 - i] = keydata[i + 10];
+  std::array<u8, 6> skey;
+  for (int i = 0; i != 6; ++i)
+    skey[5 - i] = key_data[i + 10];
 
-  for (idx = 0; idx < 7; ++idx)
+  int idx;
+  for (idx = 0; idx != 7; ++idx)
   {
-    GenerateKey(rand, idx, testkey);
-    if (0 == memcmp(testkey, skey, 6))
+    if (GenerateKey(rand, idx) == skey)
       break;
   }
+
   // default case is idx = 7 which is valid (homebrew uses it for the 0x17 case)
 
-  GenerateTables(rand, skey, idx, ft, sb);
+  // TODO: An idx of 7 does not make sense.
+  // The tables won't match the key and it will cause overflow with 3rd-party key gen.
+  // Which homebrew makes use of this so testing can be done?
+
+  return GenerateTables(rand, skey, idx);
+}
+
+std::array<u8, 6> KeyGen1stParty::GenerateKey(const std::array<u8, 10>& rand, const u8 idx) const
+{
+  constexpr u8 ans_tbl[7][6] = {
+      {0xa8, 0x77, 0xa6, 0xe0, 0xf7, 0x43}, {0x5a, 0x35, 0x85, 0xe2, 0x72, 0x97},
+      {0x8f, 0xb7, 0x1a, 0x62, 0x87, 0x38}, {0x0d, 0x67, 0xc7, 0xbe, 0x4f, 0x3e},
+      {0x20, 0x76, 0x37, 0x8f, 0x68, 0xb7}, {0xa9, 0x26, 0x3f, 0x2b, 0x10, 0xe3},
+      {0x30, 0x7e, 0x90, 0x0e, 0x85, 0x0a},
+  };
+
+  auto& ans = ans_tbl[idx];
+
+  u8 t0[10];
```

Comment:
```suggestion
  std::array<u8, 10> t0;
```
which allows using `.size()` below to only code the size in one place